src/pybind/mgr/volumes/fs/operations/clone_index.py

```diff
@@ -64,6 +64,14 @@ def get_oldest_clone_entry(self, exclude=[]):
                 sink_path = self.fs.readlink(os.path.join(self.path, min_ctime_entry[0]), CloneIndex.PATH_MAX)
                 return (min_ctime_entry[0], sink_path[:linklen])
             except cephfs.Error as e:
```

Comment:
Ignore the above comment, I see that's been handled in the second commit 9ef9e17dfcbf6df76823f812441992bb5e87ed93 
Isn't it a better idea to squash both commit into one ? What's the advantage of splitting this into two especially when the same author is involved? Won't the commit history of the file become clumsy ?

Reply:
I don't mind squashing, I kept them as two separate commits because  first one fixes the bug and second is actually an improvement that increases scope of try-except block.

Reply:
I suggest to squash them into single commit.

Reply:
Okay, in that case it will be hard to point in future that exactly what fixed the bug -- extending scope try-except or catching ObjectNotFound or combination of both. Right now it's clear that only first one was sufficient to catch it.

Reply:
Now I see the actual traceback in the tracker.

```
ile "/home/rishabh/repos/ceph/mgr-vol-clone-stats/src/pybind/mgr/volumes/fs/async_job.py", line 52, in run vol_job = self.async_job.get_job()
^^^^^^^^^^^^^^^^^^^^^^^^
ile "/home/rishabh/repos/ceph/mgr-vol-clone-stats/src/pybind/mgr/volumes/fs/async_job.py", line 195, in get_job (ret, job) = self.get_next_job(volname, running_jobs)
^^^^^^^^^^^^^^^.
^^^^^^^^^^^^^^^
ile "/home/rishabh/repos/ceph/mgr-vol-clone-stats/src/pybind/mgr/volumes/fs/async_cloner.py", line 402, in get_next_job return get_next_clone_entry(self.fs_client, self.vc.volspec, volname, running_jobs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ile "/home/rishabh/repos/ceph/mgr-vol-clone-stats/src/pybind/mgr/volumes/fs/async_cloner.py", line 35, in get_next_clone_entry job = clone_index.get_oldest_clone_entry(running_jobs)
ΑΛΛΛΛΛΛΛ
ΜΑΛΛΙ
ile "/home/rishabh/repos/ceph/mgr-vol-clone-stats/src/pybind/mgr/volumes/fs/operations/clone_index.py", line 80, in get_oldest_clone_en
st = self.fs.lstat(dpath)
ΑΛΛΛΛΛΛ
ile "cephfs.pyx", line 1997, in cephfs.LibCephFS.lstat
ile "cephfs.pyx", line 1978, in cephfs.LibCephFS.stat
hfs.ObjectNotFound: error in stat: /volumes/_index/clone/3efa5bf7-115c-4f27-8b83-8fecc16a690a: No such file or directory [Errno 2]
```

It is very clearly logged that the `ile "cephfs.pyx", line 1978, in cephfs.LibCephFS.stat` called from ` get_oldest_clone_entry
st = self.fs.lstat(dpath) `is raising the exception cephfs.ObjectNotFound.

The only instance of lstat is at line no. 57 (https://github.com/ceph/ceph/commit/60821ecfd32cb9332713108e1ad6ec3a1ae8a8c9#diff-32b8a30d8ea6ea9fd5b4753c03bfa92c14dd72aab44cdb973856448514d0b320R57)

I think the whole confusion started because of `cephfs.ObjectNotFound` in the traceback. The general coding standard/approach followed in mgr/volumes is to catch `cephfs.Error` when a call is made to libcephfs. Please note that `cephfs.Error' is the parent exception class of all the exceptions raised by python binding call to libcephfs (cephfs.pyx). So `cephfs.Error` would catch any exception from libcephfs and the code relies on `ret` value which is actual linux errno to decide on how to treat the exception.

So there is only one bug here, i.e., the exception from `self.fs.lstat` is not being caught. 
The `self.fs.readlink` exception is being caught correctly and the code added to catch `cephfs.ObjectNotFound` will be dead code, since `cephfs.Error` would catch it before.

Reply:
> . Please note that cephfs.Error is the parent exception class of all the exceptions(including cephfs.ObjectNotFound) 

Copying from `src/pybind/cephfs/cephfs.pyx` -

```
class Error(Exception):
    def get_error_code(self):
        return 1
```

```
class ObjectNotFound(OSError):
    pass
```

Reply:
Yes and `OSError` has inherited `Error` class

```
class OSError(Error): 
    def __init__(self, errno, strerror):
        super(OSError, self).__init__(errno, strerror)
        self.errno = errno
        self.strerror = "%s: %s" % (strerror, os.strerror(errno))
    
    def __str__(self):
        return '{} [Errno {}]'.format(self.strerror, self.errno)
    
    def get_error_code(self):
        return self.errno

```

Reply:
Yes. I've made the change yesterday itself. :)